### **Predictive IT Infrastructure Analytics**

#### **Data Transformation & Feature Engineering (Transformación de Datos e Ingeniería de Características)**

##### **Objetivo del Notebook**

##### El objetivo de este notebook es preparar y transformar los datos del dataset **"IT System Performance & Resource Metrics"**, garantizando su calidad y generando nuevas variables que faciliten el análisis posterior y el desarrollo de modelos predictivos.

##### Durante esta etapa se realizarán procesos de validación de datos, transformación de variables y creación de características derivadas (Feature Engineering) a partir de las métricas operativas del sistema. Estas nuevas variables permitirán representar de forma más interpretable distintos niveles de utilización y comportamiento de la infraestructura IT, proporcionando información adicional para las fases de análisis exploratorio y modelado predictivo.

##### (The objective of this notebook is to prepare and transform the data contained in the **"IT System Performance & Resource Metrics"** dataset, ensuring data quality and generating new features to support further analysis and predictive modeling.

##### This stage includes data validation, variable transformation, and feature engineering processes based on system operational metrics. The newly created features aim to provide a more interpretable representation of infrastructure behavior and resource utilization, offering additional information for subsequent exploratory analysis and predictive modeling tasks.)


### **Fase 4. Transformación de Datos**
### **1. Importación de librerias:**

In [1]:
# Importación de librerias
import pandas as pd # Librería para trabajar con datos en forma de tablas
import numpy as np # Librería para operaciones numéricas
import os # Permite trabajar con rutas y carpetas del sistema

### **2. Carga del DataSet:**

In [2]:
# Carga del dataset
df = pd.read_csv("../data/raw/Big_data_dataset.csv")

# Mostrar las primeras filas del DataFrame
df.head()

,cpu_utilization,memory_usage,disk_io,network_latency,process_count,thread_count,context_switches,cache_miss_rate,temperature,power_consumption,uptime,status
0,40.581311,43.627674,36.769917,127.990769,646,3230,1500,0.065837,82.782403,255.012964,715.803096,0
1,95.317859,39.962089,10.041088,92.399198,626,3130,243,0.123481,59.424540,81.204822,86.158314,0
2,74.539424,25.853852,17.985345,192.935206,101,303,229,0.025459,90.973363,210.685067,545.755982,0
3,61.872556,64.654000,33.500751,44.576712,52,156,1574,0.178884,37.344280,90.398182,541.918137,0
4,19.821771,52.896174,24.622378,117.983427,126,378,1164,0.167508,40.203942,87.099085,113.660688,0


### **3. Creación de etiquetas descriptivas:**

In [3]:
# Normalización de las variables numéricas
df["status_label"] = df["status"].map({
    0: "Normal",
    1: "Incident"
})
# Imprimir la cantidad de registros por cada categoría en la columna "status_label"
print(df["status_label"].value_counts())

status_label
Normal      9903
Incident      97
Name: count, dtype: int64


##### Se creó la variable `status_label` para reemplazar los valores numéricos de la variable objetivo por etiquetas más descriptivas (`Normal` e `Incident`), facilitando la interpretación de los resultados.

##### La distribución obtenida muestra 9.903 registros en estado normal y 97 registros clasificados como incidentes, evidenciando un fuerte desbalance entre ambas clases.

##### Este comportamiento es habitual en entornos de monitorización IT, donde los eventos de fallo son poco frecuentes en comparación con el funcionamiento normal del sistema.

### **4. Categorización del Uso de CPU:**

In [15]:
# Clasificación de la variable "cpu_utilization" en niveles
df["cpu_level"] = pd.cut(
    df["cpu_utilization"], # Valores de uso de CPU que serán clasificados
    bins=[0, 40, 70, 100], # Definimos los rangos de clasificación para el uso de CPU
    labels=["Low", "Medium", "High"] # Asignamos una etiqueta descriptiva a cada rango
)
# Imprimir la cantidad de registros por cada categoría en la columna "cpu_level"
print(df["cpu_level"].value_counts())

cpu_level
Low       3738
Medium    3212
High      3050
Name: count, dtype: int64


##### Se categorizó la utilización de CPU en tres niveles (`Low`, `Medium` y `High`) con el objetivo de facilitar el análisis e interpretación del comportamiento del sistema.

##### La distribución obtenida muestra 3.738 registros con uso bajo, 3.212 con uso medio y 3.050 con uso alto, reflejando una representación equilibrada de los distintos niveles de carga del procesador.

##### Estos resultados permiten analizar posteriormente si determinados niveles de utilización de CPU presentan una mayor relación con la aparición de incidentes o estados anómalos del sistema.

### **5. Categorización del uso de Memoria:**

In [5]:
# Clasificación de la variable "memory_level" en niveles
df["memory_level"] = pd.cut(
    df["memory_usage"],  # Valores de uso de memoria que serán clasificados
    bins=[0, 50, 80, np.inf], # Definición de los rangos para cada nivel de uso de memoria
    labels=["Low", "Medium", "High"] # Asignación una etiqueta descriptiva a cada rango
)
# Imprimimos la cantidad de registros por cada categoría en la columna "memory_level"
print(df["memory_level"].value_counts())

memory_level
Low       4385
Medium    3347
High      2268
Name: count, dtype: int64


##### Se categorizó el uso de memoria en tres niveles (`Low`, `Medium` y `High`) para facilitar la interpretación de los distintos estados de utilización de este recurso dentro de la infraestructura IT.

##### La distribución obtenida muestra 4.385 registros con uso bajo, 3.347 con uso medio y 2.268 con uso alto, indicando una mayor concentración de observaciones en niveles reducidos de consumo de memoria.

##### Esta clasificación permite evaluar posteriormente si los niveles elevados de utilización de memoria presentan alguna relación con la aparición de incidentes o situaciones de degradación del sistema.

### **6. Categorización de latencia:**

In [6]:
# Clasificación de la variable "network_latency" en niveles
df["latency_level"] = pd.cut(
    df["network_latency"],  # Valores de latencia de red 
    bins=[0, 50, 100, np.inf], # Definición los Rangos de clasificación
    labels=["Low", "Medium", "High"] # Etiquetas asignadas a cada rango
)
# Imprimimos la cantidad de registros por cada categoría en la columna "latency_level"
print(df["latency_level"].value_counts())

latency_level
High      5027
Low       2491
Medium    2482
Name: count, dtype: int64


##### Se categorizó la latencia de red en tres niveles (`Low`, `Medium` y `High`) con el fin de simplificar el análisis del rendimiento de las comunicaciones dentro de la infraestructura IT.

##### La distribución obtenida muestra 5.027 registros con latencia alta, 2.491 con latencia baja y 2.482 con latencia media, observándose una mayor concentración de observaciones en niveles elevados de latencia.

##### Esta clasificación facilita la identificación de posibles patrones entre el comportamiento de la red y la aparición de incidentes, permitiendo evaluar el impacto de la latencia en la estabilidad del sistema.

### **7. Categorización de temperatura:**

In [7]:
# Clasificación de la variable "temperature" en niveles
df["temperature_level"] = pd.cut(
    df["temperature"], # Valores de ltemperatura 
    bins=[0, 50, 70, np.inf], # Definición los Rangos de clasificación
    labels=["Normal", "Warning", "Critical"] # Etiquetas asignadas a cada rango
)
# Imprimir la cantidad de registros por cada categoría en la columna "temperature_level"
print(df["temperature_level"].value_counts())

temperature_level
Critical    3795
Normal      3104
Warning     3101
Name: count, dtype: int64


##### Se categorizó la temperatura del sistema en tres niveles (`Normal`, `Warning` y `Critical`) para facilitar la identificación de posibles situaciones de riesgo relacionadas con el sobrecalentamiento de la infraestructura IT.

##### La distribución obtenida muestra 3.104 registros en estado normal, 3.101 en nivel de advertencia y 3.795 en estado crítico, observándose una ligera predominancia de temperaturas elevadas dentro del conjunto de datos.

##### Esta clasificación permite analizar posteriormente si los niveles de temperatura más altos presentan una mayor asociación con la aparición de incidentes, degradación del rendimiento o posibles fallos operativos del sistema.

### **8. Normalización de variables:**

In [8]:
# Vector de nombres de columnas numéricas
columnas_numericas = [
    "cpu_utilization",
    "memory_usage",
    "network_latency",
    "temperature",
    "power_consumption"
]

# Normalización de las variables numéricas
for col in columnas_numericas:
    df[col + "_norm"] = (
        (df[col] - df[col].min())
        / (df[col].max() - df[col].min())
    )
    
    # Imprimir los valores normalizados de cada columna
    print("\n")
    print(f"Valores normalizados de {col}:")  
    print(df[col + "_norm"].head())



Valores normalizados de cpu_utilization:
0    0.374639
1    0.950982
2    0.732198
3    0.598823
4    0.156053
Name: cpu_utilization_norm, dtype: float64


Valores normalizados de memory_usage:
0    0.373570
1    0.332832
2    0.176037
3    0.607250
4    0.476577
Name: memory_usage_norm, dtype: float64


Valores normalizados de network_latency:
0    0.638277
1    0.459386
2    0.964702
3    0.219020
4    0.587978
Name: network_latency_norm, dtype: float64


Valores normalizados de temperature:
0    0.812102
1    0.452671
2    0.938145
3    0.112899
4    0.156903
Name: temperature_norm, dtype: float64


Valores normalizados de power_consumption:
0    0.820130
1    0.124683
2    0.642764
3    0.161467
4    0.148267
Name: power_consumption_norm, dtype: float64


##### Se aplicó un proceso de normalización a variables clave del sistema, como la utilización de CPU, uso de memoria, latencia de red, temperatura y consumo energético, transformando sus valores a una escala común entre 0 y 1.

##### Los resultados muestran que todas las variables fueron correctamente escaladas, permitiendo comparar métricas originalmente expresadas en unidades y rangos diferentes sin que ninguna tenga un peso desproporcionado en los análisis posteriores.

##### Esta transformación resulta especialmente útil para la creación de indicadores compuestos y para el entrenamiento de modelos de Machine Learning, ya que mejora la comparabilidad entre variables y favorece la estabilidad de los algoritmos predictivos.

### **Fase 5. Feature Engineering**
### **9. Creación del indicador de presión del sistema:**

In [9]:
# Categorizar el score de presión del sistema en niveles
df["system_pressure_score"] = (
    df["cpu_utilization_norm"] +
    df["memory_usage_norm"] +
    df["network_latency_norm"] +
    df["temperature_norm"] +
    df["power_consumption_norm"]
) / 5

# Imprimir los valores de presión del sistema
print("Valores del score de presión del sistema:")
print(df["system_pressure_score"].head())

Valores del score de presión del sistema:
0    0.603744
1    0.464111
2    0.690769
3    0.339892
4    0.305156
Name: system_pressure_score, dtype: float64


##### Se creó el indicador `system_pressure_score`, calculado a partir de la combinación de variables normalizadas como utilización de CPU, uso de memoria, latencia de red, temperatura y consumo energético.

##### Los valores obtenidos oscilan entre 0 y 1, donde puntuaciones más elevadas representan una mayor presión operativa sobre la infraestructura IT y puntuaciones más bajas reflejan condiciones de funcionamiento más estables.

##### Este indicador permite resumir múltiples métricas técnicas en una única variable, facilitando el análisis global del estado del sistema y la identificación de posibles situaciones de sobrecarga o riesgo operativo.

### **10. Clasificación del nivel de presión:**

In [10]:
# Clasificación del score de presión del sistema en niveles
df["pressure_level"] = pd.cut(
    df["system_pressure_score"],
    bins=[0, 0.33, 0.66, 1],
    labels=["Low", "Medium", "High"],
    include_lowest=True
)
# Imprimir la cantidad de registros por cada categoría en la columna "pressure_level"
print("Valores del nivel de presión del sistema:")
print(df["pressure_level"].value_counts())

Valores del nivel de presión del sistema:
pressure_level
Medium    7934
High      1101
Low        965
Name: count, dtype: int64


##### Se clasificó el indicador `system_pressure_score` en tres niveles de presión (`Low`, `Medium` y `High`) con el objetivo de facilitar la interpretación del estado operativo general de la infraestructura IT.

##### La distribución obtenida muestra 965 registros con presión baja, 7.934 con presión media y 1.101 con presión alta, evidenciando que la mayoría de los sistemas operan bajo condiciones de carga moderada.

##### Esta categorización permite identificar de forma más sencilla situaciones potenciales de sobrecarga y analizar posteriormente si los niveles elevados de presión presentan una mayor relación con la aparición de incidentes o degradaciones del rendimiento.

### **11. Exportación:**

In [11]:
# Creación de ruta de guardado de DataFrame transformado en un nuevo archivo CSV
ruta_dataframe_tr = "../data/processed"
os.makedirs(ruta_dataframe_tr, exist_ok=True)

# Guardar el DataFrame transformado en un nuevo archivo CSV
df.to_csv(
    "../data/processed/Big_data_dataset_transformado.csv",
    index=False
)
# Imprimir mensaje de confirmación
print("Dataset transformado guardado como 'Big_data_dataset_transformado.csv'")

Dataset transformado guardado como 'Big_data_dataset_transformado.csv'


### **12. Validación de las Transformaciones Aplicadas:**

In [12]:
# Cargar datasets
df_original = pd.read_csv("../data/raw/Big_data_dataset.csv")
df_transformado = pd.read_csv("../data/processed/Big_data_dataset_transformado.csv")

# Imprimir las dimensiones de los datasets original y transformado
print("="*50)
print("DIMENSIONES")
print("="*50)

print("Original:", df_original.shape)
print("Transformado:", df_transformado.shape)

print("\n")

print("="*50)
print("COLUMNAS AÑADIDAS")
print("="*50)

# Identificar las columnas nuevas en el dataset transformado
columnas_nuevas = set(df_transformado.columns) - set(df_original.columns)

# Imprimir las columnas nuevas
for col in sorted(columnas_nuevas):
    print(col)

print("\n")

print("="*70)
print("TIPOS DE DATOS")
print("="*70)

# Crea un DataFrame para comparar los tipos de datos entre el dataset original y el transformado
comparacion_tipos = pd.DataFrame({
    "Original": df_original.dtypes,
    "Transformado": df_transformado.dtypes
})

# Comparar los tipos de datos y agregar una columna que indique si la variable es nueva o original
comparacion_tipos = (
    comparacion_tipos
    .reset_index()
    .rename(columns={"index": "Variable"})
)
# Agregar una columna que indique si la variable es nueva o original
comparacion_tipos["Estado"] = comparacion_tipos["Original"].isna().map({
    True: "**Nueva variable**",
    False: "Variable original"
})
# Comparacion de tipos de datos entre el dataset original y el transformado
comparacion_tipos

DIMENSIONES
Original: (10000, 12)
Transformado: (10000, 24)


COLUMNAS AÑADIDAS
cpu_level
cpu_utilization_norm
latency_level
memory_level
memory_usage_norm
network_latency_norm
power_consumption_norm
pressure_level
status_label
system_pressure_score
temperature_level
temperature_norm


TIPOS DE DATOS


,Variable,Original,Transformado,Estado
0,cache_miss_rate,float64,float64,Variable original
1,context_switches,int64,int64,Variable original
2,cpu_level,NaN,str,**Nueva variable**
3,cpu_utilization,float64,float64,Variable original
4,cpu_utilization_norm,NaN,float64,**Nueva variable**
5,disk_io,float64,float64,Variable original
6,latency_level,NaN,str,**Nueva variable**
7,memory_level,NaN,str,**Nueva variable**
8,memory_usage,float64,float64,Variable original
9,memory_usage_norm,NaN,float64,**Nueva variable**


##### Se realizó una validación final de las transformaciones aplicadas para verificar la integridad y consistencia del dataset resultante.

##### Los resultados muestran que el conjunto de datos mantuvo sus 10.000 registros originales, mientras que el número de columnas aumentó de 12 a 24 gracias a la incorporación de nuevas variables derivadas, categorías descriptivas e indicadores calculados.

##### Entre las nuevas variables creadas destacan las categorizaciones de CPU, memoria, latencia, temperatura y presión del sistema, así como las versiones normalizadas de las métricas principales y el indicador `system_pressure_score`.

##### La validación confirma que todas las transformaciones se aplicaron correctamente, conservando las variables originales e incorporando nuevas características que enriquecen el análisis y mejoran la preparación del dataset para las fases posteriores de exploración y modelado predictivo.

In [13]:
# Verificar si hay columnas eliminadas, si el resultado es mayor a 0, significa que hay columnas eliminadas
columnas_eliminadas = set(df_original.columns) - set(df_transformado.columns)

# Imprimir el número de columnas eliminadas y sus nombres
print("Número de columnas eliminadas:", len(columnas_eliminadas))
print(columnas_eliminadas)

Número de columnas eliminadas: 0
set()


##### Se verificó la existencia de columnas eliminadas durante el proceso de transformación y preparación de datos.

##### Los resultados muestran que no fue necesario eliminar ninguna variable del dataset, manteniéndose las 12 columnas originales junto con las nuevas características generadas durante el proceso de transformación.

##### Esta validación confirma que todas las variables originales fueron consideradas útiles para el análisis y que el enriquecimiento del dataset se realizó mediante la creación de nuevas características sin pérdida de información inicial.

### **12.1. Justificación de las Transformaciones Aplicadas:**

Las transformaciones realizadas sobre el dataset tienen como finalidad enriquecer la información disponible y facilitar la interpretación de los datos durante las fases de análisis exploratorio y modelado predictivo.

Dado que el conjunto de datos original no presentaba valores nulos ni registros duplicados, no fue necesario aplicar técnicas de limpieza correctiva. Por este motivo, las acciones realizadas se centraron en la creación de nuevas variables derivadas y en la preparación de los datos para análisis avanzados.

Tras comparar el dataset original y el dataset transformado, se verificó que no fue necesario eliminar registros ni columnas, ya que durante la fase de exploración no se detectaron valores nulos, registros duplicados ni inconsistencias significativas en los datos.

El proceso de transformación se centró en la creación de nuevas variables derivadas orientadas al análisis y al modelado predictivo. Como resultado, el número de registros se mantuvo constante (10.000 filas), mientras que el número de atributos aumentó de 12 a 23 columnas debido a la incorporación de variables categóricas, métricas normalizadas e indicadores compuestos.

Las principales transformaciones fueron:

- Creación de etiquetas descriptivas para la variable objetivo (`status_label`), facilitando la interpretación de los estados del sistema.
- Categorización de métricas críticas como utilización de CPU, uso de memoria, latencia de red y temperatura, permitiendo identificar niveles de riesgo de forma más intuitiva.
- Normalización de variables numéricas para homogeneizar escalas y facilitar futuros procesos de modelado.
- Construcción de un indicador compuesto (`system_pressure_score`) capaz de resumir el estado general de carga y presión del sistema.
- Clasificación de los sistemas según niveles de presión (`pressure_level`) para simplificar análisis comparativos y segmentaciones.

Estas transformaciones no modifican los datos originales, sino que añaden nuevas características analíticas que mejoran la capacidad de extracción de conocimiento y permiten una mejor comprensión del comportamiento de la infraestructura IT.

### **12.2. Resultado del Proceso de Transformación:**

Las transformaciones realizadas fueron de carácter no destructivo, centrándose en la creación de nuevas variables derivadas orientadas al análisis y al modelado predictivo. Como resultado, el número de registros se mantuvo constante (10.000 filas), mientras que el número de atributos aumentó de 12 a 23 columnas debido a la incorporación de variables categóricas, métricas normalizadas e indicadores compuestos.

Este enfoque permite conservar la integridad del dataset original al mismo tiempo que se enriquece la información disponible para las fases posteriores del proyecto.

### **12.3. Normalización de Variables Numéricas:**

La normalización es una técnica de transformación de datos que permite convertir variables con diferentes escalas a un rango común entre 0 y 1.

En este proyecto, variables como **la utilización de CPU**, **el uso de memoria**, **la latencia de red**, **la temperatura** y **el consumo energético** presentan magnitudes distintas, lo que puede dificultar su comparación directa y afectar el rendimiento de futuros modelos predictivos.

Para evitar este problema, se aplicó la técnica **Min-Max Scaling**, que transforma cada valor utilizando la siguiente expresión:

Donde:

- \(x\) representa el valor original.
- \(x_{min}\) es el valor mínimo de la variable.
- \(x_{max}\) es el valor máximo de la variable.
- \(x_{norm}\) es el valor normalizado.

Tras la transformación:

- El valor mínimo de cada variable pasa a ser 0.
- El valor máximo pasa a ser 1.
- El resto de los valores quedan distribuidos proporcionalmente dentro de ese rango.

Las variables normalizadas generadas durante esta fase fueron:

- cpu_utilization_norm
- memory_usage_norm
- network_latency_norm
- temperature_norm
- power_consumption_norm

Estas nuevas variables serán utilizadas posteriormente para la construcción del indicador compuesto **system_pressure_score**, así como en posibles modelos de **Machine Learning** donde resulta conveniente trabajar con variables en escalas homogéneas.

La normalización no modifica la información original del dataset, sino que facilita la comparación entre métricas heterogéneas y permite combinar distintas variables en indicadores agregados sin que aquellas con valores numéricamente más altos tengan una influencia desproporcionada sobre los resultados.

### **13. Tareas completadas en este notebook:**
- Carga del dataset "IT System Performance & Resource Metrics".
- Validación de la estructura general del conjunto de datos.
- Verificación de la calidad de los datos obtenida durante la fase de exploración.
- Validación de los tipos de datos de las variables originales.
- Creación de variables categóricas para facilitar la interpretación de métricas operativas:
  - cpu_level
  - memory_level
  - latency_level
  - temperature_level
- Normalización de variables numéricas relevantes:
  - cpu_utilization_norm
  - memory_usage_norm
  - network_latency_norm
  - temperature_norm
  - power_consumption_norm
- Creación del indicador compuesto **system_pressure_score** para resumir el estado general de la infraestructura.
- Clasificación del nivel de presión del sistema mediante la variable **pressure_level**.
- Creación de la variable descriptiva **status_label** para facilitar la interpretación de la variable objetivo.
- Comparación entre el **dataset original** y el **dataset transformado**.
- Verificación de que no se eliminaron registros ni columnas del conjunto de datos original.
- Exportación del dataset transformado y guardado en la carpeta **/data/processed/** para las fases posteriores de análisis y modelado.